mixer/adapter/prometheus/prometheus.go

```diff
@@ -209,29 +241,77 @@ func (h *handler) HandleMetric(_ context.Context, vals []*metric.Instance) error
 				result = multierror.Append(result, fmt.Errorf("could not get value for metric %s: %v", val.Name, err))
 				continue
 			}
-			vec.With(promLabels(val.Dimensions)).Set(amt)
+			pl := promLabels(val.Dimensions)
+			if h.labelsCache != nil {
+				h.labelsCache.Set(key(val.Name, "gauge", pl), &cacheEntry{vec, pl})
+			}
+			vec.With(pl).Set(amt)
 		case config.COUNTER:
 			vec := collector.(*prometheus.CounterVec)
 			amt, err := promValue(val.Value)
 			if err != nil {
 				result = multierror.Append(result, fmt.Errorf("could not get value for metric %s: %v", val.Name, err))
 				continue
 			}
-			vec.With(promLabels(val.Dimensions)).Add(amt)
+			pl := promLabels(val.Dimensions)
+			if h.labelsCache != nil {
+				h.labelsCache.Set(key(val.Name, "counter", pl), &cacheEntry{vec, pl})
+			}
+			vec.With(pl).Add(amt)
 		case config.DISTRIBUTION:
 			vec := collector.(*prometheus.HistogramVec)
 			amt, err := promValue(val.Value)
 			if err != nil {
 				result = multierror.Append(result, fmt.Errorf("could not get value for metric %s: %v", val.Name, err))
 				continue
 			}
-			vec.With(promLabels(val.Dimensions)).Observe(amt)
+			pl := promLabels(val.Dimensions)
+			if h.labelsCache != nil {
+				h.labelsCache.Set(key(val.Name, "distribution", pl), &cacheEntry{vec, pl})
+			}
+			vec.With(pl).Observe(amt)
 		}
 	}
 
 	return result.ErrorOrNil()
 }
 
+func key(name, kind string, labels prometheus.Labels) uint64 {
+	keys := make([]string, 0, len(labels))
+	// ensure stable order
+	for k := range labels {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	buf := pool.GetBuffer()
```

Comment:
This looks like an expensive operation, with memory allocations and sorts.

1. Instead of using Buffer we should do multiple `h.Write()` s
2 Can we use the fact the we cannot re-dimension configuration without reloading the handler.
  Since (If) that is the case, then we can store sorted labels at construction time and then just use them here, instead of alloc and sort every time.

Reply:
I wrote a little benchmark test to eval the impact of not using `pool.GetBuffer` (after refactoring out the sort bits).

Test code:
```
func BenchmarkKey(b *testing.B) {
	pl := prometheus.Labels(map[string]string{
		"a":      "14.5",
		"c":      "hello",
		"f":      "cannonball",
		"b":      "kubernetes",
		"g":      "istio",
		"d":      "system",
		"magic":  "15ms",
		"lots":   "tons",
		"of":     "0.0001",
		"fun":    "isn't it?",
		"test":   "production",
		"joy":    "sadness",
		"allocs": "zero"})

	keys := make([]string, 0, len(pl))
	// ensure stable order
	for k := range pl {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	b.ReportAllocs()
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		key("testMetric", "gauge", pl, keys)
	}
}
```

Impl with `Buffer`:
```
func key(name, kind string, labels prometheus.Labels, sortedLabelKeys []string) uint64 {
	buf := pool.GetBuffer()
	buf.WriteString(name + ":" + kind)
	for _, k := range sortedLabelKeys {
		buf.WriteString(k)         // nolint: gas
		buf.WriteString("=")       // nolint: gas
		buf.WriteString(labels[k]) // nolint: gas
		buf.WriteString(";")       // nolint: gas
	}
	h := fnv.New64()
	h.Write(buf.Bytes())
	pool.PutBuffer(buf)
	return h.Sum64()
}
```

Impl without `Buffer`:
```
func key(name, kind string, labels prometheus.Labels, sortedLabelKeys []string) uint64 {
	h := fnv.New64()
	h.Write([]byte(name + ":" + kind))
	for _, k := range sortedLabelKeys {
	        h.Write([]byte(k + "=" + labels[k] + ";"))
	}
	return h.Sum64()
}
```

Benchmark results:

With `Buffer`:
`BenchmarkKey-8   	  500000	      3637 ns/op	       8 B/op	       1 allocs/op`

Without `Buffer`:
`BenchmarkKey-8   	 1000000	      2256 ns/op	     208 B/op	      15 allocs/op`

So, without `Buffer`, the `key` operation is faster, but results in more allocations and bytes per operation.

Thoughts?

Reply:
Oh, and if I optimize the `Buffer` impl to use:

```
for _, k := range sortedLabelKeys {
		buf.WriteString(k + "=" + labels[k] + ";") // nolint: gas
}
```

I get results like:

`BenchmarkKey-8   	 1000000	      1880 ns/op	       8 B/op	       1 allocs/op`